src/messages/MOSDForceRecovery.h

```diff
@@ -69,12 +69,36 @@ struct MOSDForceRecovery : public Message {
 
   void encode_payload(uint64_t features) {
     using ceph::encode;
+    if (!HAVE_FEATURE(features, SERVER_MIMIC)) {
+      vector<pg_t> pgs;
+      for (auto pgid : forced_pgs) {
+	pgs.push_back(pgid.pgid);
+      }
+      encode(fsid, payload);
+      encode(pgs, payload);
+      encode(options, payload);
+      return;
+    }
+    header.version = HEAD_VERSION;
+    header.compat_version = COMPAT_VERSION;
     encode(fsid, payload);
     encode(forced_pgs, payload);
     encode(options, payload);
   }
   void decode_payload() {
     bufferlist::iterator p = payload.begin();
+    if (header.version == 1) {
+      vector<pg_t> pgs;
+      decode(fsid, p);
+      decode(pgs, p);
+      decode(options, p);
+      for (auto pg : pgs) {
+	// note: this only works with replicated pools.  if a pre-mimic mon
+	// tries to force a mimic+ osd on an ec pool it will not work.
```

Comment:
Sorry, I don't really understand how the forced recovery happens. Did it not previously work in EC pools at all? Or is this a regression?

Reply:
Mmmm, that's probably okay but will need to include documentation. People occasionally get in to a lot of trouble during upgrades (or upgrade because of trouble!) and we need to make it clear that they can't do this.

...actually, we need to figure out a way to block this sensibly. Probably throw a clog error?